allennlp/common/testing/test_case.py

```diff
@@ -18,8 +18,22 @@ class AllenNlpTestCase:
     PROJECT_ROOT = (pathlib.Path(__file__).parent / ".." / ".." / "..").resolve()
     MODULE_ROOT = PROJECT_ROOT / "allennlp"
     TOOLS_ROOT = MODULE_ROOT / "tools"
-    TESTS_ROOT = PROJECT_ROOT / "tests"
-    FIXTURES_ROOT = PROJECT_ROOT / "test_fixtures"
+    # to run test suite with finished package, which does not contain
+    # tests & fixtures, we must be able to look them up somewhere else
+    PROJECT_ROOT_FALLBACK = (
+        # users wanting to run test suite for installed package
+        pathlib.Path(os.environ["ALLENNLP_SRC_DIR"])
+        if "ALLENNLP_SRC_DIR" in os.environ
+        else (
+            # fallback for conda packaging
+            pathlib.Path(os.environ["SRC_DIR"])
+            if "CONDA_BUILD" in os.environ
+            # stay in-tree
+            else PROJECT_ROOT
+        )
+    )
+    TESTS_ROOT = PROJECT_ROOT_FALLBACK / "tests"
+    FIXTURES_ROOT = PROJECT_ROOT_FALLBACK / "test_fixtures"
```

Comment:
Would it make more sense to call this `TESTS_ROOT` instead?

```suggestion
    TESTS_ROOT = (
        # users wanting to run test suite for installed package
        pathlib.Path(os.environ["ALLENNLP_SRC_DIR"])
        if "ALLENNLP_SRC_DIR" in os.environ
        else (
            # fallback for conda packaging
            pathlib.Path(os.environ["SRC_DIR"])
            if "CONDA_BUILD" in os.environ
            # stay in-tree
            else PROJECT_ROOT
        )
    )
    TESTS_ROOT = TESTS_ROOT / "tests"
    FIXTURES_ROOT = TESTS_ROOT / "test_fixtures"
```

I guess what I'm asking is, does conda put ALL source files in this directory (tests and library), or just tests? If it's all files, then I'd rather override `PROJECT_ROOT` itself.

Reply:
This is less of a conda-specific issue and more about that the installed package (somewhere on the `$PATH`, not in the current folder) does not contain the tests.

So running the tests with in stalled package requires a source checkout to run against. Sure, one could conceivably _just_ put the tests (and README, see hunk below) in that folder, but that's a really pathological setup. IMO a full source checkout is the only thing that makes sense, which is what conda bases itself off of (though, if you want to get down to the nitty-gritty, actually only the tests are ["piped through"](https://github.com/conda-forge/allennlp-feedstock/blob/master/recipe/meta.yaml#L83-L87) to the test environment; but no-one would ever do this by hand, so from general principle and or the `ALLENNLP_SRC_DIR` case, we should assume it's always a full checkout).

Does that make sense?